// ng-flash-messages/ng-flash-messages.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/platform-browser'), require('@angular/platform-browser/animations'), require('@angular/animations')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@angular/platform-browser', '@angular/platform-browser/animations', '@angular/animations'], factory) :
	(factory((global['ng-flash-messages'] = {}),global.core,global.common,global.platformBrowser,global.animations,global.animations$1));
}(this, (function (exports,core,common,platformBrowser,animations,animations$1) { 'use strict';

var NgFlashMessageService = (function () {
    function NgFlashMessageService() {
        this.showFlashMessageEvent = new core.EventEmitter();
    }
    /**
     * @param {?} flashMessage
     * @return {?}
     */
    NgFlashMessageService.prototype.showFlashMessage = function (flashMessage) {
        this.showFlashMessageEvent.emit(flashMessage);
    };
    return NgFlashMessageService;
}());
NgFlashMessageService.decorators = [
    { type: core.Injectable },
];
/**
 * @nocollapse
 */
NgFlashMessageService.ctorParameters = function () { return []; };

var FlashMessage = (function () {
    function FlashMessage() {
    }
    return FlashMessage;
}());

var NgFlashMessageComponent = (function () {
    /**
     * @param {?} ngFlashMessageService
     */
    function NgFlashMessageComponent(ngFlashMessageService) {
        this.ngFlashMessageService = ngFlashMessageService;
        this.alertVisibility = 'invisible';
    }
    /**
     * @return {?}
     */
    NgFlashMessageComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.flashMessage = new FlashMessage;
        this.ngFlashMessageService.showFlashMessageEvent.subscribe(function (flashMessage) {
            _this.alertVisibility = 'visible';
            flashMessage.type = flashMessage.type ? flashMessage.type : 'info';
            flashMessage.timeout = flashMessage.timeout === undefined ? 2000 : flashMessage.timeout;
            _this.flashMessage = flashMessage;
            if (flashMessage.timeout) {
                setTimeout(function () {
                    _this.alertVisibility = 'invisible';
                }, flashMessage.timeout);
            }
        });
    };
    /**
     * @param {?} evt
     * @return {?}
     */
    NgFlashMessageComponent.prototype.onBtnCloseClick = function (evt) {
        evt.preventDefault();
        this.alertVisibility = 'invisible';
    };
    return NgFlashMessageComponent;
}());
NgFlashMessageComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'ng-flash-message',
                template: "<div [@alertState]=\"alertVisibility\" class=\"alert alert-{{ flashMessage.type }}\" role=\"alert\"> <a href=\"#\" *ngIf=\"flashMessage.dismissible\" (click)=\"onBtnCloseClick($event)\" class=\"close\">&times;</a> <span *ngIf=\"flashMessage.messages?.length == 1\" [innerHTML]=\"flashMessage.messages[0]\"> </span> <span *ngIf=\"flashMessage.messages?.length > 1\"> <p *ngFor=\"let message of flashMessage.messages\" [innerHTML]=\"message\"> </p> </span> </div>",
                styles: [".alert { padding: 15px; margin-bottom: 20px; border: 1px solid transparent; border-radius: 4px; } .alert h4 { margin-top: 0; color: inherit; } .alert .alert-link { font-weight: bold; } .alert > p, .alert > ul { margin-bottom: 0; } .alert > p + p { margin-top: 5px; } .alert-dismissable, .alert-dismissible { padding-right: 35px; } .alert-dismissable .close, .alert-dismissible .close { position: relative; top: -2px; right: -21px; color: inherit; } .alert-success { background-color: #dff0d8; border-color: #d6e9c6; color: #3c763d; } .alert-success hr { border-top-color: #c9e2b3; } .alert-success .alert-link { color: #2b542c; } .alert-info { background-color: #d9edf7; border-color: #bce8f1; color: #31708f; } .alert-info hr { border-top-color: #a6e1ec; } .alert-info .alert-link { color: #245269; } .alert-warning { background-color: #fcf8e3; border-color: #faebcc; color: #8a6d3b; } .alert-warning hr { border-top-color: #f7e1b5; } .alert-warning .alert-link { color: #66512c; } .alert-danger { background-color: #f2dede; border-color: #ebccd1; color: #a94442; } .alert-danger hr { border-top-color: #e4b9c0; } .alert-danger .alert-link { color: #843534; } .alert a.close { float: right; text-decoration: none !important; color: #777777; font-weight: bold; } .alert a.close:hover { color: #333333; } "],
                animations: [
                    animations$1.trigger('alertState', [
                        animations$1.state('visible', animations$1.style({
                            opacity: 1,
                            display: 'block'
                        })),
                        animations$1.state('invisible', animations$1.style({
                            opacity: 0,
                            display: 'none'
                        })),
                        animations$1.transition('visible => invisible', animations$1.animate('500ms ease-in')),
                        animations$1.transition('invisible => visible', animations$1.animate('0ms linear'))
                    ])
                ]
            },] },
];
/**
 * @nocollapse
 */
NgFlashMessageComponent.ctorParameters = function () { return [
    { type: NgFlashMessageService, },
]; };

var NgFlashMessagesModule = (function () {
    function NgFlashMessagesModule() {
    }
    /**
     * @return {?}
     */
    NgFlashMessagesModule.forRoot = function () {
        return {
            ngModule: NgFlashMessagesModule,
            providers: [NgFlashMessageService]
        };
    };
    return NgFlashMessagesModule;
}());
NgFlashMessagesModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [
                    common.CommonModule,
                    platformBrowser.BrowserModule,
                    animations.BrowserAnimationsModule
                ],
                declarations: [
                    NgFlashMessageComponent
                ],
                providers: [
                    NgFlashMessageService
                ],
                exports: [
                    NgFlashMessageComponent
                ]
            },] },
];
/**
 * @nocollapse
 */
NgFlashMessagesModule.ctorParameters = function () { return []; };

exports.NgFlashMessagesModule = NgFlashMessagesModule;
exports.NgFlashMessageService = NgFlashMessageService;

Object.defineProperty(exports, '__esModule', { value: true });

})));
